File: 83_merge/merge.cpp
/*
 * @Date: 2021-04-05 10:26:55
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-04-05 10:31:16
 * @FilePath: \algorithm\83_merge\merge.cpp
 * @Description: file content
 */

#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
  int p1 = m - 1, p2 = n - 1;
  int tail = m + n - 1;
  int cur;
  while (p1 >= 0 || p2 >= 0) {
    if (p1 == -1) {
      cur = nums2[p2--];
    } else if (p2 == -1) {
      cur = nums1[p1--];
    } else if (nums1[p1] > nums2[p2]) {
      cur = nums1[p1--];
    } else {
      cur = nums2[p2--];
    }
    nums1[tail--] = cur;
  }
}

int main() {
  {
    vector<int> nums1 = {1, 2, 3, 0, 0, 0};
    int m = 3;
    vector<int> nums2 = {2, 5, 6};
    int n = 3;
    merge(nums1, m, nums2, n);
    for (int i : nums1) cout << i << ", ";
  }
  cout << endl;
  {
    vector<int> nums1 = {1};
    int m = 1;
    vector<int> nums2 = {};
    int n = 0;
    merge(nums1, m, nums2, n);
    for (int i : nums1) cout << i << ", ";
  }
}

File: rust/88_merge/merge.rs
/*
 * @Date: 2023-08-13
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-08-13
 * @FilePath: /algorithm/rust/88_merge/merge.rs
 */

struct Solution;
impl Solution {
    pub fn merge(nums1: &mut Vec<i32>, m: i32, nums2: &mut Vec<i32>, n: i32) {
        let mut i1 = m as usize;
        let mut i2 = n as usize;
        let mut i = (m + n) as usize;

        // nums2 处理完就可以结束了
        while i2 > 0 {
            if i1 == 0 || nums2[i2 - 1] > nums1[i1 - 1] {
                nums1[i - 1] = nums2[i2 - 1];
                i2 -= 1;
            } else {
                nums1[i - 1] = nums1[i1 - 1];
                i1 -= 1;
            }
            i -= 1;
        }
    }
}

fn main() {
    let tests = vec![
        (
            vec![1, 2, 3, 0, 0, 0],
            3,
            vec![2, 5, 6],
            3,
            vec![1, 2, 2, 3, 5, 6],
        ),
        (vec![1], 1, vec![], 0, vec![1]),
        (vec![0], 0, vec![1], 1, vec![1]),
    ];

    for (mut nums1, m, mut nums2, n, ans) in tests {
        Solution::merge(&mut nums1, m, &mut nums2, n);
        assert_eq!(nums1, ans);
    }
}
